aspects/aspects_test.go

```diff
@@ -1521,3 +1506,196 @@ func (*aspectSuite) TestSetPreCheckMultipleAlternativeTypesHappy(c *C) {
+		},
+	}, aspects.NewJSONSchema())
+	c.Assert(err, IsNil)
+
+	asp := aspectBundle.Aspect("foo")
+	c.Assert(asp, NotNil)
+
+	err = asp.Set(databag, "foo", map[string]interface{}{
+		"a": map[string]interface{}{"a": map[string]interface{}{
+			"c": "value",
+			"d": "other",
+		}},
+		"b": map[string]interface{}{"a": map[string]interface{}{
+			"e": "value",
+			"f": "other",
+		}},
+	})
+	c.Assert(err, IsNil)
+
+	data, err := asp.Get(databag, "foo")
+	c.Assert(err, IsNil)
+	c.Assert(data, DeepEquals, map[string]interface{}{
+		"a": map[string]interface{}{"a": map[string]interface{}{
+			"c": "value",
+			"d": "other",
+		}},
+		"b": map[string]interface{}{"a": map[string]interface{}{
+			"e": "value",
+			"f": "other",
+		}},
+	})
+}
+
+func (s *aspectSuite) TestUnsetUnmatchedPlaceholder(c *C) {
+	databag := aspects.NewJSONDataBag()
+	aspectBundle, err := aspects.NewBundle("acc", "bundle", map[string]interface{}{
+		"foo": []interface{}{
+			map[string]interface{}{"request": "foo.{bar}", "storage": "foo.{bar}"},
+		},
+	}, aspects.NewJSONSchema())
+	c.Assert(err, IsNil)
+
+	asp := aspectBundle.Aspect("foo")
+	c.Assert(asp, NotNil)
+
+	err = asp.Set(databag, "foo", map[string]interface{}{
+		"bar": "value",
+		"baz": "other",
+	})
+	c.Assert(err, IsNil)
+
+	err = asp.Set(databag, "foo", nil)
+	c.Assert(err, ErrorMatches, `cannot set "foo" in aspect acc/bundle/foo: cannot unset with unmatched placeholders`)
+}
+
+func (s *aspectSuite) TestGetValuesThroughPaths(c *C) {
+	type testcase struct {
+		path     string
+		suffix   []string
+		value    interface{}
+		expected map[string]interface{}
+		err      string
+	}
+
+	tcs := []testcase{
+		{
+			path:     "foo.bar",
+			suffix:   nil,
+			value:    "value",
+			expected: map[string]interface{}{"foo.bar": "value"},
+		},
+		{
+			path:     "foo.{bar}",
+			suffix:   []string{"{bar}"},
+			value:    map[string]interface{}{"a": "value", "b": "other"},
+			expected: map[string]interface{}{"foo.a": "value", "foo.b": "other"},
+		},
+		{
+			path:   "foo.{bar}.baz",
+			suffix: []string{"{bar}", "baz"},
+			value: map[string]interface{}{
+				"a": map[string]interface{}{"baz": "value"},
+				"b": map[string]interface{}{"baz": "other"},
+			},
+			expected: map[string]interface{}{"foo.a.baz": "value", "foo.b.baz": "other"},
+		},
+		{
+			path:   "foo.{bar}.{baz}.last",
+			suffix: []string{"{bar}", "{baz}"},
+			value: map[string]interface{}{
+				"a": map[string]interface{}{"b": "value"},
+				"c": map[string]interface{}{"d": "other"},
+			},
+			expected: map[string]interface{}{"foo.a.b.last": "value", "foo.c.d.last": "other"},
+		},
+
+		{
+			path:   "foo.{bar}",
+			suffix: []string{"{bar}", "baz"},
+			value: map[string]interface{}{
+				"a": map[string]interface{}{"baz": "value", "ignore": 1},
+				"b": map[string]interface{}{"baz": "other", "ignore": 1},
+			},
+			expected: map[string]interface{}{"foo.a": "value", "foo.b": "other"},
```

Comment:
@pedronis you briefly mentioned this issue before, do you think having unused keys here should be an error instead? I thought it made sense to just ignore unused keys because it seems more flexible but of course it may let user mistakes go unnoticed (if the user thinks those values are going to be set and they're not, for instance) and it would be harder if we were to remove this later. However, this allows us to have one rule that sets "baz" and another one that sets "ignore", and set them both with one call to set. It would also be more flexible if we wanted to read a top-level value like `foo`, edit a subset of that data and then set the whole thing back, including things that won't be set. If we don't permit sub-paths to be ignored, the value would have to be "cleaned up" before. wdyt?

Reply:
> However, this allows us to have one rule that sets "baz" and another one that sets "ignore", 

can you make an example of this?  you mean rules with unrelated prefixes? or you mean rule that we would use at the same time in a multi-set?,   also what happens if placeholders are not involved? do we get an error? we need to be consistent, we cannot have a behavior with placeholder that is different if everything is specified

Reply:
Let's say we have two rules:
```
rules:
  -
    request: {snap}.version
    storage: snaps.{snap}.version
  -
    request: {snap}.status
    storage: snaps.{snap}.status
    access: read
```
And some data stored for snap `foo`. If we ignore unused keys, we could do something like
```
$ snap get a/b/c foo
// data={"foo": { "version": 1, "status": "enabled" }}

// increment version

$ snap set a/b/c foo data
```

If we let Set ignore data it doesn't need, then this works. If we don't, the user needs to clean up the data before setting.
A different but related case would be:

```
rules:
  -
    request: {snap}.version
    storage: snaps.{snap}.version
  -
    request: {snap}.status
    storage: snaps.{snap}.status

# the same but the status rule can also be used to write
```

Then reading "foo", making a change to both the version and the status and setting back. This would also work because the write to each path would ignore the other field and we write to both storage paths in one operation. In this case we could enforce a restriction that every path value needs to be used at least once globally, if we wanted to

Reply:
but if we exploded the first into:

snap set a/b/c foo.version=1
snap set a/b/c foo.status="enabled"

the 2nd one would error? so the operation with data should also error, no?

>  In this case we could enforce a restriction that every path value needs to be used at least once globally, if we wanted to

that seems to be going in the right direction

Reply:
Implement this in the latest commit